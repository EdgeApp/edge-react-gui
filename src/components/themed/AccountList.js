// @flow

import { type EdgeUserInfo } from 'edge-core-js'
import * as React from 'react'
import { Pressable, ScrollView, TouchableHighlight, View } from 'react-native'
import Animated, { Easing, useAnimatedStyle, useSharedValue, withTiming } from 'react-native-reanimated'
import Feather from 'react-native-vector-icons/Feather'
import MaterialIcon from 'react-native-vector-icons/MaterialIcons'
import { sprintf } from 'sprintf-js'

import { deleteLocalAccount } from '../../actions/AccountActions.js'
import { logoutRequest } from '../../actions/LoginActions.js'
import { Fontello } from '../../assets/vector'
import s from '../../locales/strings'
import { useEffect, useState } from '../../types/reactHooks'
import { useDispatch, useSelector } from '../../types/reactRedux'
import { ButtonsModal } from '../modals/ButtonsModal.js'
import { Airship } from '../services/AirshipInstance.js'
import { type Theme, cacheStyles, useTheme } from '../services/ThemeContext'
import { DividerLine } from '../themed/DividerLine'
import { EdgeText } from '../themed/EdgeText'

type Props = {
  onPress: () => void,
  isOpen: boolean,
  duration: number
}

export function AccountList(props: Props) {
  const { isOpen, onPress, duration = 500 } = props
  const activeUsername = useSelector(state => state.core.account.username)
  const context = useSelector(state => state.core.context)
  const dispatch = useDispatch()
  const theme = useTheme()
  const styles = getStyles(theme)

  // Maintain the list of usernames:
  const [usernames, setUsernames] = useState(arrangeUsers(context.localUsers, activeUsername))
  useEffect(() => context.watch('localUsers', localUsers => setUsernames(arrangeUsers(localUsers, activeUsername))), [activeUsername, context])

  const handleDelete = (username: string) => {
    Airship.show(bridge => (
      <ButtonsModal
        bridge={bridge}
        title={s.strings.delete_account_header}
        message={sprintf(s.strings.delete_username_account, username)}
        buttons={{
          ok: {
            label: s.strings.string_delete,
            async onPress() {
              await dispatch(deleteLocalAccount(username))
              height.value -= styles.row.height
              return true
            }
          },
          cancel: { label: s.strings.string_cancel }
        }}
      />
    ))
  }

  const handleSwitchAccount = (username: string) => {
    dispatch(logoutRequest(username))
  }

  const listHeight = styles.row.height * usernames.length + theme.rem(1)
  const height = useSharedValue(isOpen ? listHeight : 0)
  useEffect(() => {
    height.value = isOpen ? listHeight : 0
  }, [height, isOpen, listHeight])

  const animatedStyle = useAnimatedStyle(() => {
    return {
      height: withTiming(height.value, {
        duration: duration,
        easing: Easing.out(Easing.exp)
      })
    }
  })

  return (
    <View>
      <Pressable onPress={onPress}>
        <View style={styles.header}>
          <Fontello name="edge.logo" style={styles.iconUser} size={theme.rem(1.5)} color={theme.mainMenuIcon} />
          <EdgeText style={styles.text}>{activeUsername}</EdgeText>
          <Feather name="chevron-down" color={theme.mainMenuIcon} size={theme.rem(1.5)} />
        </View>
      </Pressable>
      <DividerLine marginRem={[1, -2, 0, 0]} />
      <Animated.View style={[styles.root, animatedStyle]}>
        <ScrollView>
          {usernames.map((username: string) => (
            <View key={username} style={styles.row}>
              <TouchableHighlight onPress={() => handleSwitchAccount(username)}>
                <EdgeText style={styles.text}>{username}</EdgeText>
              </TouchableHighlight>
              <TouchableHighlight onPress={() => handleDelete(username)}>
                <View>
                  <MaterialIcon size={theme.rem(1.5)} name="close" color={theme.mainMenuIcon} />
                </View>
              </TouchableHighlight>
            </View>
          ))}
        </ScrollView>
      </Animated.View>
    </View>
  )
}

/**
 * Given a list of users from the core,
 * remove the given user, then organize the 3 most recent users,
 * followed by the rest in alphabetical order.
 */
function arrangeUsers(localUsers: EdgeUserInfo[], activeUsername: string): string[] {
  // Sort the users according to their last login date:
  const usernames = localUsers
    .filter(info => info.username !== activeUsername)
    .sort((a, b) => {
      const { lastLogin: aDate = new Date(0) } = a
      const { lastLogin: bDate = new Date(0) } = b
      return aDate.valueOf() - bDate.valueOf()
    })
    .map(info => info.username)

  // Sort everything after the last 3 entries alphabetically:
  const oldUsernames = usernames.slice(3).sort((a: string, b: string) => {
    const stringA = a.toUpperCase()
    const stringB = b.toUpperCase()
    if (stringA < stringB) return -1
    if (stringA > stringB) return 1
    return 0
  })

  return [...usernames.slice(0, 3), ...oldUsernames]
}

const getStyles = cacheStyles((theme: Theme) => ({
  row: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    height: theme.rem(2.5),
    marginLeft: theme.rem(2)
  },
  text: {
    marginRight: 'auto',
    fontFamily: theme.fontFaceBold
  },
  header: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    height: theme.rem(2)
  },
  iconUser: {
    marginRight: theme.rem(1.5)
  }
}))
