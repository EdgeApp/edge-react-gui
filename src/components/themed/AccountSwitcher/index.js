// @flow

export * from './helpers'
export { default as SwitcherHeader } from './SwitcherHeader'
export { default as SwitcherList } from './SwitcherList'
