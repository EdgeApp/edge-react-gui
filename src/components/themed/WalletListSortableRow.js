// @flow

import { bns } from 'biggystring'
import { type EdgeCurrencyWallet, type EdgeDenomination } from 'edge-core-js'
import * as React from 'react'
import { ActivityIndicator, TouchableOpacity, View } from 'react-native'
import FastImage from 'react-native-fast-image'
import Ionicon from 'react-native-vector-icons/Ionicons'

import { formatNumberInput } from '../../locales/intl.js'
import { getDisplayDenominationFromState } from '../../selectors/DenominationSelectors.js'
import { calculateWalletFiatBalanceWithoutState } from '../../selectors/WalletSelectors.js'
import { connect } from '../../types/reactRedux.js'
import { type GuiExchangeRates } from '../../types/types.js'
import { getCurrencyIcon } from '../../util/CurrencyInfoHelpers.js'
import { getWalletName } from '../../util/CurrencyWalletHelpers.js'
import { DECIMAL_PRECISION, decimalOrZero, getFiatSymbol, truncateDecimals } from '../../util/utils'
import { type Theme, type ThemeProps, cacheStyles, withTheme } from '../services/ThemeContext.js'
import { EdgeText } from './EdgeText.js'

type OwnProps = {
  wallet?: EdgeCurrencyWallet
}

type StateProps = {
  exchangeRates: GuiExchangeRates,
  showBalance: boolean,
  walletFiatSymbol: string | null
}

type DispatchProps = {
  getDisplayDenomination: (pluginId: string, currencyCode: string) => EdgeDenomination
}

type Props = OwnProps & StateProps & ThemeProps & DispatchProps

class WalletListSortableRowComponent extends React.PureComponent<Props> {
  render() {
    const { wallet, walletFiatSymbol, exchangeRates, showBalance, theme, getDisplayDenomination } = this.props
    // $FlowFixMe react-native-sortable-listview sneakily injects this prop:
    const { sortHandlers } = this.props
    const styles = getStyles(theme)

    if (wallet == null) {
      return (
        <View style={styles.container}>
          <TouchableOpacity activeOpacity={0.95} underlayColor={theme.underlayColor} {...sortHandlers}>
            <View style={[styles.rowContainer, styles.loaderContainer]}>
              <ActivityIndicator color={theme.primaryText} size="small" />
            </View>
          </TouchableOpacity>
        </View>
      )
    }

    const displayDenomination = getDisplayDenomination(wallet.currencyInfo.pluginId, wallet.currencyInfo.currencyCode)
    const multiplier = displayDenomination.multiplier
    const name = getWalletName(wallet)
    const symbol = displayDenomination.symbol
    const currencyCode = wallet.currencyInfo.currencyCode
    const { symbolImageDarkMono } = getCurrencyIcon(currencyCode)

    const balance = wallet.balances[currencyCode] ?? '0'
    const preliminaryCryptoAmount = truncateDecimals(bns.div(balance, multiplier, DECIMAL_PRECISION), 6)
    const finalCryptoAmount = formatNumberInput(decimalOrZero(preliminaryCryptoAmount, 6)) // make it show zero if infinitesimal number
    const finalCryptoAmountString = showBalance ? `${symbol || ''} ${finalCryptoAmount}` : ''
    const fiatBalance = calculateWalletFiatBalanceWithoutState(wallet, currencyCode, exchangeRates)
    const fiatBalanceFormat = fiatBalance && parseFloat(fiatBalance) > 0.000001 ? fiatBalance : 0
    const fiatBalanceSymbol = showBalance && walletFiatSymbol ? walletFiatSymbol : ''
    const fiatBalanceString = showBalance ? fiatBalanceFormat : ''

    return (
      <View style={styles.container}>
        <TouchableOpacity {...sortHandlers}>
          <View style={styles.rowContainer}>
            <View style={styles.iconContainer}>
              <Ionicon name="ios-menu" size={theme.rem(1.25)} color={theme.icon} />
            </View>
            <View style={styles.iconContainer}>
              {symbolImageDarkMono && <FastImage style={styles.icon} source={{ uri: symbolImageDarkMono }} resizeMode="cover" />}
            </View>
            <View style={styles.detailsContainer}>
              <View style={styles.detailsRow}>
                <EdgeText style={styles.detailsCurrency}>{currencyCode}</EdgeText>
                <EdgeText style={styles.detailsValue}>{finalCryptoAmountString}</EdgeText>
              </View>
              <View style={styles.detailsRow}>
                <EdgeText style={styles.detailsName}>{name}</EdgeText>
                <EdgeText style={styles.detailsFiat}>{fiatBalanceSymbol + fiatBalanceString}</EdgeText>
              </View>
            </View>
          </View>
        </TouchableOpacity>
      </View>
    )
  }
}

const getStyles = cacheStyles((theme: Theme) => ({
  container: {
    paddingHorizontal: theme.rem(1)
  },
  rowContainer: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    height: theme.rem(4.25),
    backgroundColor: theme.tileBackground
  },
  loaderContainer: {
    justifyContent: 'center',
    alignItems: 'center'
  },
  iconContainer: {
    alignItems: 'center',
    justifyContent: 'center',
    width: theme.rem(1.25),
    marginRight: theme.rem(1)
  },
  icon: {
    width: theme.rem(2),
    height: theme.rem(2),
    resizeMode: 'contain'
  },
  detailsContainer: {
    flex: 1,
    flexDirection: 'column'
  },
  detailsRow: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center'
  },
  detailsCurrency: {
    flex: 1
  },
  detailsValue: {
    textAlign: 'right'
  },
  detailsName: {
    flex: 1,
    fontSize: theme.rem(0.75),
    color: theme.secondaryText
  },
  detailsFiat: {
    fontSize: theme.rem(0.75),
    textAlign: 'right',
    color: theme.secondaryText
  },
  exchangeRate: {
    flex: 1,
    fontSize: theme.rem(0.75),
    textAlign: 'left'
  },
  percentage: {
    fontSize: theme.rem(0.75),
    fontFamily: theme.fontFaceMedium
  },
  divider: {
    height: theme.rem(1 / 16),
    borderColor: theme.lineDivider,
    borderBottomWidth: theme.rem(1 / 16),
    marginVertical: theme.rem(0.5)
  }
}))

export const WalletListSortableRow = connect<StateProps, DispatchProps, OwnProps>(
  (state, ownProps) => ({
    showBalance: state.ui.settings.isAccountBalanceVisible,
    exchangeRates: state.exchangeRates,
    walletFiatSymbol: ownProps.wallet ? getFiatSymbol(ownProps.wallet.fiatCurrencyCode) : null
  }),
  dispatch => ({
    getDisplayDenomination(pluginId: string, currencyCode: string) {
      return dispatch(getDisplayDenominationFromState(pluginId, currencyCode))
    }
  })
)(withTheme(WalletListSortableRowComponent))
