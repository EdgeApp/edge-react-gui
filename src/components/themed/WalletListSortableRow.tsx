import { div, gt } from 'biggystring'
import { EdgeCurrencyWallet } from 'edge-core-js'
import * as React from 'react'
import { ActivityIndicator, TouchableOpacity, View } from 'react-native'
import Ionicon from 'react-native-vector-icons/Ionicons'

import { FIAT_PRECISION, getSymbolFromCurrency } from '../../constants/WalletAndCurrencyConstants'
import { formatNumber, formatNumberInput } from '../../locales/intl'
import { getDisplayDenominationFromState, getExchangeDenomination } from '../../selectors/DenominationSelectors'
import { calculateFiatBalance } from '../../selectors/WalletSelectors'
import { useDispatch, useSelector } from '../../types/reactRedux'
import { getWalletName } from '../../util/CurrencyWalletHelpers'
import { DECIMAL_PRECISION, decimalOrZero, truncateDecimals } from '../../util/utils'
import { CryptoIcon } from '../icons/CryptoIcon'
import { cacheStyles, Theme, useTheme } from '../services/ThemeContext'
import { EdgeText } from './EdgeText'

interface Props {
  wallet?: EdgeCurrencyWallet
  onDrag: () => void
}

function WalletListSortableRowComponent(props: Props) {
  const { wallet, onDrag } = props

  const theme = useTheme()
  const styles = getStyles(theme)
  const dispatch = useDispatch()

  const showBalance = useSelector(state => state.ui.settings.isAccountBalanceVisible)
  const exchangeRates = useSelector(state => state.exchangeRates)
  const exchangeDenomination = useSelector(state =>
    wallet == null ? null : getExchangeDenomination(state, wallet.currencyInfo.pluginId, wallet.currencyInfo.currencyCode)
  )

  if (wallet == null || exchangeDenomination == null) {
    return (
      <TouchableOpacity style={styles.container} activeOpacity={0.95} onLongPress={onDrag}>
        <View style={[styles.rowContainer, styles.loaderContainer]}>
          <ActivityIndicator color={theme.primaryText} size="small" />
        </View>
      </TouchableOpacity>
    )
  }

  const { currencyCode, pluginId } = wallet.currencyInfo
  const walletFiatSymbol = getSymbolFromCurrency(wallet.fiatCurrencyCode)
  const displayDenomination = dispatch(getDisplayDenominationFromState(pluginId, currencyCode))
  const multiplier = displayDenomination.multiplier
  const name = getWalletName(wallet)
  const symbol = displayDenomination.symbol

  const balance = wallet.balances[currencyCode] ?? '0'
  const preliminaryCryptoAmount = truncateDecimals(div(balance, multiplier, DECIMAL_PRECISION))
  const finalCryptoAmount = formatNumberInput(decimalOrZero(preliminaryCryptoAmount, 6)) // make it show zero if infinitesimal number
  const finalCryptoAmountString = showBalance ? `${symbol || ''} ${finalCryptoAmount}` : ''
  const fiatBalance = calculateFiatBalance(wallet, exchangeDenomination, exchangeRates)
  const fiatBalanceFormat = fiatBalance && gt(fiatBalance, '0.000001') ? fiatBalance : 0
  const fiatBalanceSymbol = showBalance && walletFiatSymbol ? walletFiatSymbol : ''
  const fiatBalanceString = showBalance ? formatNumber(fiatBalanceFormat, { toFixed: FIAT_PRECISION }) : ''

  return (
    <TouchableOpacity style={styles.container} onLongPress={onDrag}>
      <View style={styles.rowContainer}>
        <View style={styles.iconContainer}>
          <Ionicon name="ios-menu" size={theme.rem(1.25)} color={theme.icon} />
        </View>
        <View style={styles.iconContainer}>
          <CryptoIcon pluginId={wallet.currencyInfo.pluginId} walletId={wallet.id} />
        </View>
        <View style={styles.detailsContainer}>
          <View style={styles.detailsRow}>
            <EdgeText style={styles.detailsCurrency}>{currencyCode}</EdgeText>
            <EdgeText style={styles.detailsValue}>{finalCryptoAmountString}</EdgeText>
          </View>
          <View style={styles.detailsRow}>
            <EdgeText style={styles.detailsName}>{name}</EdgeText>
            <EdgeText style={styles.detailsFiat}>{fiatBalanceSymbol + fiatBalanceString}</EdgeText>
          </View>
        </View>
      </View>
    </TouchableOpacity>
  )
}

const getStyles = cacheStyles((theme: Theme) => ({
  container: {
    paddingHorizontal: theme.rem(1)
  },
  rowContainer: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    height: theme.rem(4.25),
    backgroundColor: theme.tileBackground
  },
  loaderContainer: {
    justifyContent: 'center',
    alignItems: 'center'
  },
  iconContainer: {
    alignItems: 'center',
    justifyContent: 'center',
    width: theme.rem(1.25),
    marginRight: theme.rem(1)
  },
  icon: {
    width: theme.rem(2),
    height: theme.rem(2),
    resizeMode: 'contain'
  },
  detailsContainer: {
    flex: 1,
    flexDirection: 'column'
  },
  detailsRow: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center'
  },
  detailsCurrency: {
    flex: 1
  },
  detailsValue: {
    textAlign: 'right'
  },
  detailsName: {
    flex: 1,
    fontSize: theme.rem(0.75),
    color: theme.secondaryText
  },
  detailsFiat: {
    fontSize: theme.rem(0.75),
    textAlign: 'right',
    color: theme.secondaryText
  },
  exchangeRate: {
    flex: 1,
    fontSize: theme.rem(0.75),
    textAlign: 'left'
  },
  percentage: {
    fontSize: theme.rem(0.75),
    fontFamily: theme.fontFaceMedium
  },
  divider: {
    height: theme.rem(1 / 16),
    borderColor: theme.lineDivider,
    borderBottomWidth: theme.rem(1 / 16),
    marginVertical: theme.rem(0.5)
  }
}))

export const WalletListSortableRow = React.memo(WalletListSortableRowComponent)
