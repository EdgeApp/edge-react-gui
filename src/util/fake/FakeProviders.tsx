import * as React from 'react'
import { Metrics, SafeAreaProvider } from 'react-native-safe-area-context'
import { Provider } from 'react-redux'
import { applyMiddleware, createStore } from 'redux'
import thunk from 'redux-thunk'

import { rootReducer, RootState } from '../../reducers/RootReducer'

type DeepPartial<T> = T extends object
  ? {
      [P in keyof T]?: DeepPartial<T[P]>
    }
  : T

export type FakeState = DeepPartial<RootState>

interface Props {
  children: React.ReactNode
  initialState?: FakeState
}

export function FakeProviders(props: Props) {
  const { children, initialState = {} } = props

  const store = React.useMemo(() => createStore(rootReducer, initialState as any, applyMiddleware(thunk)), [initialState])
  return (
    <SafeAreaProvider initialMetrics={initialMetrics}>
      <Provider store={store}>{children}</Provider>
    </SafeAreaProvider>
  )
}

// These match what our old react-native-safe-area-view library returned:
const initialMetrics: Metrics = {
  frame: { height: 1334, width: 750, x: 0, y: 0 },
  insets: { bottom: 0, left: 0, right: 0, top: 20 }
}
