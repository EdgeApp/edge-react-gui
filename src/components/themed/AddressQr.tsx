import { EdgeCurrencyWallet, EdgeTokenId } from 'edge-core-js'
import * as React from 'react'
import { View } from 'react-native'

import { useAsyncValue } from '../../hooks/useAsyncValue'
import { getCurrencyCode } from '../../util/CurrencyInfoHelpers'
import { cacheStyles, Theme, useTheme } from '../services/ThemeContext'
import { QrCode } from '../themed/QrCode'

interface Props {
  address: string
  tokenId: EdgeTokenId
  wallet: EdgeCurrencyWallet

  nativeAmount?: string
  onPress?: (encodedUri?: string) => void
}

export const AddressQr = (props: Props) => {
  const theme = useTheme()
  const styles = getStyles(theme)
  const { address, tokenId, wallet, nativeAmount, onPress = () => {} } = props

  const [encodedUri] = useAsyncValue(
    async () => await wallet.encodeUri({ publicAddress: address, currencyCode: getCurrencyCode(wallet, tokenId), nativeAmount }),
    [address, tokenId, nativeAmount, wallet]
  )

  return (
    <View style={styles.container}>
      <QrCode data={encodedUri} tokenId={tokenId} wallet={wallet} onPress={() => onPress(encodedUri)} marginRem={0} />
    </View>
  )
}

const getStyles = cacheStyles((theme: Theme) => ({
  container: {
    aspectRatio: 1,
    height: '100%',
    width: '100%',
    alignSelf: 'center'
  }
}))
