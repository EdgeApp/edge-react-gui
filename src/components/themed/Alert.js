// @flow

import * as React from 'react'
import { Platform, StyleSheet, TouchableOpacity, View } from 'react-native'
import IonIcon from 'react-native-vector-icons/Ionicons'

import { fixSides, mapSides, sidesToMargin, sidesToPadding } from '../../util/sides.js'
import { type Theme, cacheStyles, useTheme } from '../services/ThemeContext.js'
import { EdgeText } from '../themed/EdgeText'

type AlerType = 'info' | 'warning' | 'error' | 'success'

type Props = {
  type: AlerType,
  title: string,
  message?: string,
  numberOfLines?: number,
  children?: React.Node,
  marginRem?: number[] | number,
  paddingRem?: number[] | number,
  onPress?: () => void
}

type StylesOptions = { typeColor: string }

// TODO: Implement alert colors such as info and others
export const useAlertTypeColor = (type: AlerType) => {
  const theme = useTheme()

  switch (type) {
    case 'warning':
      return theme.warningText
    case 'error':
      return theme.dangerText
    default:
      return theme.primaryText
  }
}

// TODO: Implement alert icons such as info and others
export const useAlertTypeIcon = (type: AlerType, color: string, size: number, style: StyleSheet.Styles) => {
  switch (type) {
    case 'warning':
    case 'error':
      return (
        <IonIcon style={style} name={Platform.OS === 'ios' ? 'ios-information-circle-outline' : 'md-information-circle-outline'} color={color} size={size} />
      )
    default:
      return null
  }
}

export function Alert({ type, title, message, numberOfLines = 2, marginRem, paddingRem, onPress, children }: Props) {
  const theme = useTheme()
  const margin = sidesToMargin(mapSides(fixSides(marginRem, 1), theme.rem))
  const padding = sidesToPadding(mapSides(fixSides(paddingRem, 1), theme.rem))

  const typeColor = useAlertTypeColor(type)

  const styles = getStyles(theme, { typeColor })

  const icon = useAlertTypeIcon(type, typeColor, theme.rem(1.25), styles.icon)

  const result = (
    <View style={[styles.alert, margin, padding]}>
      <View style={styles.header}>
        {icon}
        <EdgeText style={styles.title}>{title}</EdgeText>
      </View>
      {message && (
        <EdgeText style={styles.message} numberOfLines={numberOfLines}>
          {message}
        </EdgeText>
      )}
      {children}
    </View>
  )

  return onPress ? <TouchableOpacity onPress={onPress}>{result}</TouchableOpacity> : result
}

const getStyles = (theme: Theme, { typeColor }: StylesOptions) =>
  cacheStyles((theme: Theme) => ({
    alert: {
      borderWidth: theme.cardBorder,
      borderColor: typeColor,
      borderRadius: theme.cardBorderRadius,
      alignSelf: 'stretch'
    },
    header: {
      display: 'flex',
      flexDirection: 'row',
      alignItems: 'center'
    },
    title: {
      fontSize: theme.rem(0.75),
      color: typeColor
    },
    icon: {
      marginRight: theme.rem(0.5)
    },
    message: {
      fontSize: theme.rem(0.75),
      color: typeColor,
      fontFamily: theme.fontFaceDefault
    }
  }))(theme)
