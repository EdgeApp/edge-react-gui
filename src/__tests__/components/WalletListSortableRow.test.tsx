import { describe, expect, it } from '@jest/globals'
import { EdgeCurrencyInfo } from 'edge-core-js'
import * as React from 'react'
import TestRenderer from 'react-test-renderer'

import { WalletListSortableRow } from '../../components/themed/WalletListSortableRow'
import { FakeProviders, FakeState } from '../../util/fake/FakeProviders'

describe('WalletListSortableRow', () => {
  it('should render with loading wallet', () => {
    const renderer = TestRenderer.create(
      <FakeProviders>
        <WalletListSortableRow wallet={undefined} onDrag={() => {}} />
      </FakeProviders>
    )
    expect(renderer.toJSON()).toMatchSnapshot()
    renderer.unmount()
  })

  it('should render with fake wallet', () => {
    const fakeCurrencyInfo: Partial<EdgeCurrencyInfo> = {
      pluginId: 'fake',
      currencyCode: 'FAKE',
      denominations: [{ name: 'FAKE', multiplier: '1' }]
    }
    const fakeState: FakeState = {
      core: {
        account: {
          currencyConfig: {
            fake: {
              allTokens: {},
              currencyInfo: fakeCurrencyInfo
            }
          },
          watch: () => () => {}
        }
      }
    }
    const fakeWallet: any = {
      currencyInfo: fakeCurrencyInfo,
      name: 'Test wallet',
      balances: {}
    }

    const renderer = TestRenderer.create(
      <FakeProviders initialState={fakeState}>
        <WalletListSortableRow wallet={fakeWallet} onDrag={() => {}} />
      </FakeProviders>
    )
    expect(renderer.toJSON()).toMatchSnapshot()
    renderer.unmount()
  })
})
