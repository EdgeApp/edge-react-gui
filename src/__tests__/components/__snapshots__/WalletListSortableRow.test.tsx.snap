// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletListSortableRow should render with fake wallet 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
      "paddingHorizontal": 22,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "rgba(255, 255, 255, 0)",
        "flexDirection": "row",
        "height": 96,
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "justifyContent": "center",
          "marginRight": 22,
          "width": 28,
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          [
            {
              "color": "#FFFFFF",
              "fontSize": 28,
            },
            undefined,
            {
              "fontFamily": "Ionicons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        ÔèΩ
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "justifyContent": "center",
          "marginRight": 22,
          "width": 28,
        }
      }
    >
      <View
        style={
          {
            "height": 45,
            "marginBottom": 0,
            "marginLeft": 0,
            "marginRight": 0,
            "marginTop": 0,
            "width": 45,
          }
        }
      >
        <View
          style={
            [
              {
                "overflow": "hidden",
              },
              {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              },
            ]
          }
        >
          <FastImageView
            resizeMode="cover"
            source={
              {
                "uri": "https://content.edge.app/currencyIconsV3/fake/fake.png",
              }
            }
            style={
              {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          />
        </View>
      </View>
    </View>
    <View
      style={
        {
          "flex": 1,
          "flexDirection": "column",
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "center",
          }
        }
      >
        <Text
          adjustsFontSizeToFit={true}
          minimumFontScale={0.65}
          numberOfLines={1}
          style={
            [
              {
                "color": "#FFFFFF",
                "fontFamily": "Quicksand-Regular",
                "fontSize": 22,
                "includeFontPadding": false,
              },
              {
                "flex": 1,
              },
              null,
            ]
          }
        >
          FAKE
        </Text>
        <Text
          adjustsFontSizeToFit={true}
          minimumFontScale={0.65}
          numberOfLines={1}
          style={
            [
              {
                "color": "#FFFFFF",
                "fontFamily": "Quicksand-Regular",
                "fontSize": 22,
                "includeFontPadding": false,
              },
              {
                "textAlign": "right",
              },
              null,
            ]
          }
        >
           0
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "center",
          }
        }
      >
        <Text
          adjustsFontSizeToFit={true}
          minimumFontScale={0.65}
          numberOfLines={1}
          style={
            [
              {
                "color": "#FFFFFF",
                "fontFamily": "Quicksand-Regular",
                "fontSize": 22,
                "includeFontPadding": false,
              },
              {
                "color": "#A4C7DF",
                "flex": 1,
                "fontSize": 17,
              },
              null,
            ]
          }
        >
          Test wallet
        </Text>
        <Text
          adjustsFontSizeToFit={true}
          minimumFontScale={0.65}
          numberOfLines={1}
          style={
            [
              {
                "color": "#FFFFFF",
                "fontFamily": "Quicksand-Regular",
                "fontSize": 22,
                "includeFontPadding": false,
              },
              {
                "color": "#A4C7DF",
                "fontSize": 17,
                "textAlign": "right",
              },
              null,
            ]
          }
        >
          0.00
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`WalletListSortableRow should render with loading wallet 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
      "paddingHorizontal": 22,
    }
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "rgba(255, 255, 255, 0)",
          "flexDirection": "row",
          "height": 96,
          "justifyContent": "center",
        },
        {
          "alignItems": "center",
          "justifyContent": "center",
        },
      ]
    }
  >
    <ActivityIndicator
      color="#FFFFFF"
      size="small"
    />
  </View>
</View>
`;
